h2o-core/src/main/java/water/H2O.java

```diff
@@ -1324,6 +1325,9 @@ public static String DEFAULT_FLOW_DIR() {
    * It is updated also when a new client appears. */
   private static HashSet<H2ONode> STATIC_H2OS = null;
 
+  /* List of all clients that ever connected to this cloud */
+  private static Map<H2ONode.H2Okey, H2ONode> CLIENTS_MAP = new ConcurrentHashMap<>();
```

Comment:
I meant Cliff's NBHM. But at the end it does not matter, Cliff's NBHM has better performance under high load, i do not expect 10^6 clients to connect and disconnect.